presto-array/src/main/java/com/facebook/presto/array/AdaptiveLongBigArray.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.array;
+
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.slice.SizeOf.sizeOfLongArray;
+
+/**
+ * This variation of BigArray is designed to expand segments up to some reasonable extend
+ * and add more segments if the maximum segment capacity is reached
+ * This implementation allows to keep the redirection table small so it does fit into L1 CPU cache
+ */
+public class AdaptiveLongBigArray
+{
+    // visible for testing
+    static final int INSTANCE_SIZE = ClassLayout.parseClass(AdaptiveLongBigArray.class).instanceSize();
+
+    // settings are constants due to efficiency considerations
+    static final int INITIAL_SEGMENT_LENGTH = 16 * 1024; // 128KB
+    static final int MAX_SEGMENT_LENGTH = 32 * 1024 * 1024; // 256MB
+    static final long MAX_SEGMENT_SIZE_IN_BYTES = sizeOfLongArray(MAX_SEGMENT_LENGTH);
+    static final int INITIAL_SEGMENTS = 10;
+    static final int SEGMENT_SHIFT = 25;
+    static final int SEGMENT_MASK = MAX_SEGMENT_LENGTH - 1;
+
+    // segments are allocated lazily in ensureCapacity
+    // The first segment is allocated initially with INITIAL_SEGMENT_LENGTH
+    // and then gradually expanded until it reaches MAX_SEGMENT_LENGTH
+    // Second and subsequent segments are directly allocated with MAX_SEGMENT_LENGTH
+    private long[][] array;
+    // number of allocated segments
+    private int segments;
+    // number of elements that currently can be stored in the container
+    private int capacity;
```

Comment:
This should be a `long` now that https://github.com/prestodb/presto/pull/15604 merged and fixed all the other classes.

Reply:
Currently this class only used by PageIndex, that does not support storing more than 2^31 elements, thus the index is still int.

Reply:
We can potentially improve it as a separate PR

Reply:
Oh woops, I didn't realize this class only used `int` for indicies and capacity. My bad!